test_engine/terraform/aws/ubuntu/main.tf

```diff
@@ -0,0 +1,221 @@
+  instance_type = var.build_engine_aws_instance_type
+
+  subnet_id = aws_subnet.build_engine_aws_public_subnet.id
+  vpc_security_group_ids = [
+    aws_security_group.build_engine_aws_secgrp.id
+  ]
+
+  root_block_device {
+    delete_on_termination = true
+    volume_size = var.build_engine_aws_instance_root_block_device_size
+  }
+
+  key_name = aws_key_pair.build_engine_aws_pair_key.key_name
+  user_data = file("${path.module}/user-data-scripts/provision.sh")
+  
+  tags = {
+    Name = "${var.build_engine_aws_instance_name}-${count.index}-${random_string.random_suffix.id}"
+    DoNotDelete	= "true"
+    Owner = "longhorn-infra"
+  }
+}
+
+resource "aws_eip" "build_engine_aws_eip_controlplane" {
+  count    = var.build_engine_aws_instance_count
+  vpc      = true
+}
+
+# Associate every EIP with controlplane instance
+resource "aws_eip_association" "build_engine_aws_eip_assoc" {
+  depends_on = [
+    aws_instance.build_engine_aws_instance,
+    aws_eip.build_engine_aws_eip_controlplane
+  ]
+
+  count    = var.build_engine_aws_instance_count
+
+  instance_id   = element(aws_instance.build_engine_aws_instance, count.index).id
+  allocation_id = element(aws_eip.build_engine_aws_eip_controlplane, count.index).id
+}
+
+# wait for docker to start on instances (for rke on amd64 only)
+resource "null_resource" "wait_for_docker_start" {
+  depends_on = [
+    aws_instance.build_engine_aws_instance,
+    aws_eip.build_engine_aws_eip_controlplane,
+    aws_eip_association.build_engine_aws_eip_assoc
+  ]
+
+  count = var.build_engine_aws_instance_count
+
+  provisioner "remote-exec" {
+    inline = ["until( systemctl is-active docker.service ); do echo \"waiting for docker to start \"; sleep 2; done"]
+
+    connection {
+      type     = "ssh"
+      user     = "ubuntu"
+      host     = element(aws_eip.build_engine_aws_eip_controlplane, count.index).public_ip
+      private_key = file(var.aws_ssh_private_key_file_path)
+    }
+  }
+}
+
+# build engine base image with different commit ID
+resource "null_resource" "build_images" {
+  depends_on = [
+    null_resource.wait_for_docker_start,
+    aws_eip.build_engine_aws_eip_controlplane,
+  ]
+
+  count = var.build_engine_aws_instance_count
+
+  provisioner "file" {
+    source      = "${var.tf_workspace}/scripts/build_images.sh"
+    destination = "/tmp/build_images.sh"
+
+    connection {
+      type     = "ssh"
+      user     = "ubuntu"
+      host     = element(aws_eip.build_engine_aws_eip_controlplane, count.index).public_ip
+      private_key = file(var.aws_ssh_private_key_file_path)
+    }
+  }
+
+  provisioner "remote-exec" {
+    inline = [
+        "sudo chmod +x /tmp/build_images.sh",
+        "sudo /tmp/build_images.sh ${var.build_engine_arch} ${var.docker_id} ${var.docker_password} ${var.docker_repo} ${var.CLIAPIVersion} ${var.CLIAPIMinVersion} ${var.ControllerAPIVersion} ${var.ControllerAPIMinVersion} ${var.DataFormatVersion} ${var.DataFormatMinVersion}"
```

Comment:
Need to ensure the account and password will not be disclosed like displaying in the jenkins job log.

Reply:
Currently in my local terminal I can not see those info
Will setup local jenkins pipe to observe jenkins log